File: 2022/day-08/day-08.c
/* --- Advent of Code 2023 - Day 8: Treetop Tree House --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 256

/* solution part-1 */
static int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* file opened successfully */
    if (fp != NULL) {
        /* read and check first line */
        if (!fgets(buffer, MAX_LENGTH - 1, fp)) {
            fprintf(stderr, "Could not read file '%s'\n", input_file);
            exit(1);
        }

        /* initialize array for map */
        int width = strlen(buffer) - 1;
        int height = width;
        char **map = (char **)malloc(sizeof(char *) * height);

        /* read map line by line */
        for (int i = 0; i < height; i++) {
            map[i] = (char *)malloc(sizeof(char) * width);
            for (int j = 0; j < width; j++) {
                map[i][j] = buffer[j];
            }
            /* get the next line and check format */
            if (!fgets(buffer, MAX_LENGTH - 1, fp) && (i < height - 1)) {
                fprintf(stderr, "Format error: height does not equal width: %d\n", i);
                exit(1);
            }
        }

        /* count visible trees */
        int visible_cnt = 0;
        int hidden = 0;

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {

                /* check if tree is visible to north */
                hidden = 0;
                for (int k = 0; k <= i - 1; k++) {
                    if (map[k][j] >= map[i][j]) {
                        hidden = 1;
                        break;
                    }
                }
                if (!hidden) {
                    visible_cnt += 1;
                    continue;
                }

                /* check if tree is visible to south */
                hidden = 0;
                for (int k = i + 1; k < height; k++) {
                    if (map[k][j] >= map[i][j]) {
                        hidden = 1;
                        break;
                    }
                }
                if (!hidden) {
                    visible_cnt += 1;
                    continue;
                }

                /* check if tree is visible to west */
                hidden = 0;
                for (int k = 0; k <= j - 1; k++) {
                    if (map[i][k] >= map[i][j]) {
                        hidden = 1;
                        break;
                    }
                }
                if (!hidden) {
                    visible_cnt += 1;
                    continue;
                }

                /* check if tree is visible to east */
                hidden = 0;
                for (int k = j + 1; k < width; k++) {
                    if (map[i][k] >= map[i][j]) {
                        hidden = 1;
                        break;
                    }
                }
                if (!hidden) {
                    visible_cnt += 1;
                    continue;
                }
            }
        }

        printf("%d\n", visible_cnt);

        /* free allocated memory */
        for (int i = 0; i < width; i++)
            free(map[i]);
        free(map);

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* solution part-2 */
static int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* file opened successfully */
    if (fp != NULL) {
        // TODO

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}

File: 2022/day-08/day-08.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2022 - Day 8: Treetop Tree House --- """

import sys


# check if tree is visible to north
def visible_to_north(i: int, j: int, input: list) -> int:
    for k in range(0, i):
        if input[k][j] >= input[i][j]:
            return 0

    return 1


# check if tree is visible to south
def visible_to_south(i: int, j: int, input: list) -> int:
    for k in range(i+1, len(input)-1):
        if input[k][j] >= input[i][j]:
            return 0

    return 1


# check if tree is visible to west
def visible_to_west(i: int, j: int, input: list) -> int:
    for k in range(0, j):
        if input[i][k] >= input[i][j]:
            return 0

    return 1


# check if tree is visible to east
def visible_to_east(i: int, j: int, input: list) -> int:
    for k in range(j+1, len(input[i])):
        if input[i][k] >= input[i][j]:
            return 0

    return 1


# viewing distance to the north?
def distance_to_north(i: int, j: int, input: list) -> int:
    distance = 0
    for k in reversed(range(0, i)):
        distance += 1
        if input[k][j] >= input[i][j]:
            break

    return distance


# viewing distance to the south?
def distance_to_south(i: int, j: int, input: list) -> int:
    distance = 0
    for k in range(i + 1, len(input) - 1):
        distance += 1
        if input[k][j] >= input[i][j]:
            break

    return distance


# viewing distance to the west?
def distance_to_west(i: int, j: int, input: list) -> int:
    distance = 0
    for k in reversed(range(0, j)):
        distance += 1
        if input[i][k] >= input[i][j]:
            break

    return distance


# viewing distance to the east?
def distance_to_east(i: int, j: int, input: list) -> int:
    distance = 0
    for k in range(j + 1, len(input[i])):
        distance += 1
        if input[i][k] >= input[i][j]:
            break

    return distance


# solution part 1
def part_1():
    with open("input.txt") as f:
        visible_cnt = 0
        input = f.read().split("\n")

        # check if tree is visible to any direction
        for i in range(len(input) - 1):
            for j in range(len(input[i])):

                # increase counter if tree is visible
                if visible_to_north(i, j, input) or visible_to_south(i, j, input) or visible_to_west(i, j, input) or visible_to_east(i, j, input):
                    visible_cnt += 1

    print(visible_cnt)


# solution part 2
def part_2():
    with open("input.txt") as f:
        max_score = 0
        input = f.read().split("\n")

        # calculate score for each tree
        for i in range(len(input) - 1):
            for j in range(len(input[i])):
                if i != 1 and j != 0 and i != 98 and j != 98:
                    score = distance_to_north(i, j, input) * distance_to_south(i, j, input) * distance_to_west(i, j, input) * distance_to_east(i, j, input)

                    # update max score
                    if score > max_score:
                        max_score = score

    print(max_score)


# main entry point
if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)
